utils/__tests__/config.test.ts

```diff
@@ -0,0 +1,199 @@
+import { Either } from "fp-ts/lib/Either";
+import { NonEmptyString } from "italia-ts-commons/lib/strings";
+import { MailerConfig } from "../config";
+
+const aMailFrom = "example@test.com";
+
+const noop = () => {};
+const expectRight = <L, R>(e: Either<L, R>, t: (r: R) => void = noop) =>
+  e.fold(
+    _ =>
+      fail(`Expecting right, received left. Value: ${JSON.stringify(e.value)}`),
+    _ => t(_)
+  );
+
+const expectLeft = <L, R>(e: Either<L, R>, t: (l: L) => void = noop) =>
+  e.fold(
+    _ => t(_),
+    _ =>
+      fail(`Expecting left, received right. Value: ${JSON.stringify(e.value)}`)
+  );
+
+describe("MailerConfig", () => {
+  it("should decode configuration for sendgrid", () => {
+    const rawConf = {
+      MAIL_FROM: aMailFrom,
+      NODE_ENV: "production",
+      SENDGRID_API_KEY: "a-sg-key"
+    };
+    const result = MailerConfig.decode(rawConf);
+
+    expectRight(result, value => {
+      expect(value.SENDGRID_API_KEY).toBe("a-sg-key");
+      expect(typeof value.MAILUP_USERNAME).toBe("undefined");
+    });
+  });
+
+  it("should decode configuration for sendgrid", () => {
+    const rawConf = {
+      MAIL_FROM: aMailFrom,
+      NODE_ENV: "production",
+      SENDGRID_API_KEY: "a-sg-key"
+    };
+    const result = MailerConfig.decode(rawConf);
+
+    expectRight(result, value => {
+      expect(value.SENDGRID_API_KEY).toBe("a-sg-key");
+      expect(typeof value.MAILUP_USERNAME).toBe("undefined");
+    });
+  });
+
+  it("should decode configuration for sendgrid even if mailup conf is passed", () => {
+    const rawConf = {
+      MAIL_FROM: aMailFrom,
+      NODE_ENV: "production",
+      SENDGRID_API_KEY: "a-sg-key",
+      MAILUP_USERNAME: "a-mu-username",
+      MAILUP_SECRET: "a-mu-secret"
+    };
+    const result = MailerConfig.decode(rawConf);
+
+    expectRight(result, value => {
+      expect(value.SENDGRID_API_KEY).toBe("a-sg-key");
+    });
+  });
+
+  it("should decode configuration for mailup", () => {
+    const rawConf = {
+      MAIL_FROM: aMailFrom,
+      NODE_ENV: "production",
+      MAILUP_USERNAME: "a-mu-username",
+      MAILUP_SECRET: "a-mu-secret"
+    };
+    const result = MailerConfig.decode(rawConf);
+
+    expectRight(result, value => {
+      expect(value.MAILUP_USERNAME).toBe("a-mu-username");
+      expect(value.MAILUP_SECRET).toBe("a-mu-secret");
+      // check types
+      const _: NonEmptyString = value.MAILUP_SECRET;
+      const __: NonEmptyString = value.MAILUP_USERNAME;
```

Comment:
This can be removed

Reply:
This is actually a canary to catch possible problems when/if we'll edit the config model. I found that several times the programmatic check doesn't ensure the correct type inference (hence, you have green tests but the project will fail to build). Not exhaustive, though.

Reply:
Sorry, I don't understand. At runtime we have ```IConfig.decode({
  ...process.env,
  isProduction: process.env.NODE_ENV === "production"
});``` to check the correct type of required values.

Reply:
If you need a type check inside the tests is better to create an explicit type
```
const MailUpConfig = t.interface({
    MAILHOG_HOSTNAME: t.undefined,
    MAILUP_SECRET: NonEmptyString,
    MAILUP_USERNAME: NonEmptyString,
    MAIL_TRANSPORTS: t.undefined,
    NODE_ENV: t.literal("production"),
    SENDGRID_API_KEY: t.undefined
})
```
And check the response. We can use these types even inside the logic with `MailUpConfig.is(config)`

Reply:
Well, the check is all about "let me se if the interface I designed actually results in the correct type once I exit the io-ts world".
In the case above, it happened that `MAILUP_USERNAME` could not be correctly narrowed to `NonEmptyString` but to `string | NonEmptyString`, although the interface was programmatically good.
Maybe I could do `NonEmptyString.is(value.MAILUP_USERNAME)`